pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,951 @@
+		endpointSliceListerSynced: endpointSliceInformer.Informer().HasSynced,
+		k8sClient:                 k8sClient,
+		bgpPeerPasswords:          make(map[string]string),
+		nodeName:                  nodeConfig.Name,
+		enabledIPv4:               networkConfig.IPv4Enabled,
+		enabledIPv6:               networkConfig.IPv6Enabled,
+		podIPv4CIDR:               nodeConfig.PodIPv4CIDR.String(),
+		podIPv6CIDR:               nodeConfig.PodIPv6CIDR.String(),
+		nodeIPv4Addr:              nodeConfig.NodeIPv4Addr.IP.String(),
+		egressEnabled:             egressEnabled,
+		newBGPServerFn: func(globalConfig *bgp.GlobalConfig) bgp.Interface {
+			return gobgp.NewGoBGPServer(globalConfig)
+		},
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "bgpPolicy"),
+	}
+	c.bgpPolicyInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addBGPPolicy,
+			UpdateFunc: c.updateBGPPolicy,
+			DeleteFunc: c.deleteBGPPolicy,
+		},
+		resyncPeriod,
+	)
+	c.serviceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addService,
+			UpdateFunc: c.updateService,
+			DeleteFunc: c.deleteService,
+		},
+		resyncPeriod,
+	)
+	c.endpointSliceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addEndpointSlice,
+			UpdateFunc: c.updateEndpointSlice,
+			DeleteFunc: c.deleteEndpointSlice,
+		},
+		resyncPeriod,
+	)
+	if c.egressEnabled {
+		c.egressInformer = egressInformer.Informer()
+		c.egressLister = egressInformer.Lister()
+		c.egressListerSynced = egressInformer.Informer().HasSynced
+		c.egressInformer.AddEventHandlerWithResyncPeriod(
+			cache.ResourceEventHandlerFuncs{
+				AddFunc:    c.addEgress,
+				UpdateFunc: c.updateEgress,
+				DeleteFunc: c.deleteEgress,
+			},
+			resyncPeriod,
+		)
+	}
+	c.nodeInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addNode,
+			UpdateFunc: c.updateNode,
+			DeleteFunc: nil,
+		},
+		resyncPeriod,
+	)
+	return c, nil
+}
+
+// watchSecretChanges uses watch API directly to watch for the changes of the Secret storing passwords of BGP
+// peers.
+func (c *Controller) watchSecretChanges(stopCh <-chan struct{}) error {
+	secretMeta := &metav1.ObjectMeta{
+		Name:      types.BGPPolicySecretName,
+		Namespace: env.GetAntreaNamespace(),
+	}
+	watcher, err := c.k8sClient.CoreV1().Secrets(secretMeta.Namespace).Watch(context.TODO(), metav1.SingleObject(*secretMeta))
+	if err != nil {
+		return fmt.Errorf("failed to create Secret watcher: %v", err)
+	}
+
+	ch := watcher.ResultChan()
+	defer watcher.Stop()
+	klog.InfoS("Starting watching Secret changes", "Secret", klog.KObj(secretMeta))
+	for {
+		select {
+		case event, ok := <-ch:
+			if !ok {
+				return nil
+			}
+			// Update BGP peer passwords.
+			klog.InfoS("Processing Secret event", "Secret", klog.KObj(secretMeta))
```

Comment:
This is a controversial topic about whether to use a variable like such cases. I think the current code is more readable.

Reply:
Right, `klog.KObj` is a util dedicated to log K8s objects in a log entry provided by klog, IMO it would be overkill to define a variable for it.

Reply:
OK, I saw there are some extra checks inside of KObj(), and it's being called 5 times, so I feel it might be worthy to define a variable, but I didn't do any comparison, so no strong opinion.